apps/devtools/app/scripts/components/MessageViewer/MessageTimeline/MessageTimeline.tsx

```diff
@@ -0,0 +1,143 @@
+  spans: [],
+  rowIndex: index,
+});
+
+export const MessageTimeline: FC<MessageTimelineProps> = ({
+  messages,
+  startedRecordingAt,
+  activeMessageId,
+  selectQuery,
+}) => {
+  /**
+   * To get this timeline to work with hovers, zooms and stuff we will needs to
+   * calculate a lot of stuff in the top level container. This would allow us to
+   * know when something is being hovered. Also the zooming of the timeline would
+   */
+  const ref = useRef<MessageTimelineMeta>({
+    startedRecordingAt,
+    currentTime: null,
+  });
+  const colors = useTimelineColors();
+  const canvas = useRef<CanvasRef<CanvasRenderingContext2D, unknown>>(null);
+  const [container, setContainer] = useState<HTMLDivElement | null>(null);
+  const [containerRect, setContainerRect] = useState<DOMRect | null>(null);
+  const { timelineSpans, rowAmount } = useMemo(() => {
+    const rows: Array<RowMeta> = [makeRowMeta(0)];
+    messages.forEach(({ id, initializedAt, timeTaken }) => {
+      let eligibleRow = rows.findIndex(
+        row => row.rowCursor + 5 < initializedAt,
+      );
+      if (eligibleRow === -1) {
+        rows.push(makeRowMeta(rows.length));
+        eligibleRow = rows.length - 1;
+      }
+      rows[eligibleRow].spans.push({
+        timeTaken: timeTaken ?? Date.now(),
+        initializedAt,
+        id,
+        rowIndex: eligibleRow,
+        color: activeMessageId === id ? colors.highlight : undefined,
+      });
+      rows[eligibleRow].rowCursor = initializedAt + (timeTaken ?? 0);
+    });
+    return {
+      timelineSpans: rows
+        .flatMap<MessageSpanType>(({ spans }) => spans.map(span => span))
+        .sort((a, b) => a.initializedAt - b.initializedAt),
+      rowAmount: rows.length,
+    };
+  }, [messages, activeMessageId]);
+
+  const onContainerRef = (containerRef: HTMLDivElement | null) => {
+    if (containerRef) {
+      setContainer(containerRef);
+    }
+  };
+
+  useEffect(() => {
+    setContainerRect(container?.getBoundingClientRect());
+  }, [container]);
+
+  useEffect(() => {
+    if (canvas.current && canvas.current.draw && canvas.current.canvas) {
+      canvas.current.draw();
+    }
+  });
+
+  const spanStartingTime = timelineSpans[0]?.initializedAt;
+  const lastSpan = timelineSpans[timelineSpans.length - 1];
+  const spanLength = timelineSpans.length;
+  const startedRecording = startedRecordingAt || spanStartingTime - 1000;
+  const currentTime = lastSpan
+    ? lastSpan.initializedAt + lastSpan.timeTaken + 1000
+    : Date.now();
+
+  ref.current.currentTime = currentTime;
+  ref.current.startedRecordingAt = startedRecording;
+  ref.current.rowAmount = Math.min(25, rowAmount);
+  ref.current.spanLength = spanLength;
+  ref.current.fullTimeline = currentTime - startedRecording;
+  ref.current.spans = timelineSpans;
+
+  return (
+    <Box
+      flex="0"
+      width="100%"
+      ref={onContainerRef}
+      borderTop="1px solid var(--border-color)"
+      borderBottom="1px solid var(--border-color)"
```

Comment:
Ooh css variables

Reply:
how does this work?

Reply:
Chakra uses CSS variables, and there is actually a little utility in here that creates a bunch of them. 

https://github.com/jointoucan/finch-graphql/blob/master/apps/devtools/app/scripts/components/Theme.tsx

It's actually a really nice way to use colors, especially when dealing with something like GraphiQL that is styled primarily with a CSS file.

Here is some docs on [CSS vars](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties).